Add error boundary to the Settings page

delawski opened the pull request:
## Summary

This PR adds the error boundary and error screen components from the Onboarding Wizard to the AMP Settings page. The error screen is also extended. It now shows not only the error message but also the stack information. Some cosmetic updates to the look and feel of the error message have been done as well. 

| Onboarding Wizard | AMP Settings |
| --- | --- |
| ![Screenshot 2021-05-31 at 14 36 11](https://user-images.githubusercontent.com/478735/120196056-b3212280-c21f-11eb-8a84-a990d03ebf72.png) | ![Screenshot 2021-05-31 at 14 36 31](https://user-images.githubusercontent.com/478735/120196066-b7e5d680-c21f-11eb-86c0-993e3e013571.png) |

| Onboarding Wizard before | Onboarding Wizard after |
| --- | --- |
| ![Screenshot 2021-05-31 at 14 50 27](https://user-images.githubusercontent.com/478735/120196162-d3e97800-c21f-11eb-9a19-cbb35b2987a8.png) | ![Screenshot 2021-05-31 at 14 50 52](https://user-images.githubusercontent.com/478735/120196176-d8159580-c21f-11eb-903e-42d43370a028.png) |

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6230

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Add WMP wizard error boundary to settings page

delawski committed:
Remove unused prop

delawski committed:
Make error boundary and error screen more generic

delawski committed:
Improve appearance of error screen

delawski committed:
Update tests and snapshot

delawski commented:
> We should wrap this with a try/catch block to show an error message with the details, including a link to submit a support topic with the associated stack trace.

@westonruter In the current iteration I haven't included the "submit a support topic" button you asked for. Should this button create a ticker on wordpress.org? If yes, is it possible to create a ticket with a predefined topic/content? I briefly searched for something like this but couldn't find anything relevant.

delawski committed:
Fix stylelint issues

delawski committed:
Minor formatting update

pierlon commented on assets/src/onboarding-wizard/index.js:
```diff
@@ -49,8 +50,13 @@ const { ajaxurl: wpAjaxUrl } = global;
-			<ErrorBoundary exitLink={ FINISH_LINK } fullScreen={ true }>
-
+			<ErrorBoundary
+				exitLink={ {
```
The prop type for `exitLink` is a `string`, but you're passing an object here.

delawski replied:
Good catch. In fact, after addressing your feedback on the `finishLink` prop, this becomes outdated.

pierlon commented on assets/src/components/error-screen/index.js:
```diff
@@ -19,27 +19,38 @@ import './style.css';
  * @param {Object} props Component props.
  * @param {Object} props.error Error object containing a message string.
- * @param {string} props.finishLink The link to return to the admin.
+ * @param {Object} props.finishLink The link to return to the admin.
```
Accepting an object for a prop named `finishLink` seems like a misnomer. Accepting the link and label as separate props seems more appropriate.

delawski replied:
I agree. I've broken down the `finishLink` object into 2 separate props in 81e29d8.

pierlon commented:
Maybe we could add a button to copy the error stack-trace as well, similar to the [one in the block editor](https://github.com/WordPress/gutenberg/blob/3c5e72e50cddb118b85220d50f1b7919c4df67c6/packages/editor/src/components/error-boundary/index.js#L77-L79).

delawski committed:
Address code review feedback

delawski committed:
Add copy to clipboard button

delawski committed:
Migrate ClipboardButton from `@wordpress/components`

delawski committed:
Add snapshot test

delawski commented:
> Maybe we could add a button to copy the error stack-trace as well, similar to the one in the block editor.

@pierlon I think it's a good idea. I've migrated the (now deprecated) [`ClipboardButton`](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/clipboard-button) from `@wordpress/components` (it's based on the `useCopyToClipboard()` custom hook that is proposed as a replacement in the [deprecation notice](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/clipboard-button/index.js#L31)).

Here's how it looks:

![Screenshot 2021-06-08 at 16 54 40](https://user-images.githubusercontent.com/478735/121223565-f6b30680-c887-11eb-9d51-f221c2dd82ad.png)

westonruter commented:
PHPUnit failures to be fixed by https://github.com/ampproject/amp-wp/pull/6366.

westonruter commented:
You can rebase now against `develop` to fix the failures.

pierlon commented:
Great work @delawski. One thing left here is to catch and handle any errors that may occur outside of the error boundary, like in https://github.com/ampproject/amp-wp/issues/6230 where an incorrect version of a dependency was loaded, causing the settings page components not to render. I think we can accomplish this by adding an `error` event listener to the page that would then render the error screen.

So maybe something like this:

```diff
diff --git a/assets/src/settings-page/index.js b/assets/src/settings-page/index.js
--- a/assets/src/settings-page/index.js	(revision 983e5e106759bb2aecd845e958494c299c850a31)
+++ b/assets/src/settings-page/index.js	(date 1623357057438)
@@ -217,6 +217,8 @@
 	const root = document.getElementById( 'amp-settings-root' );
 
 	if ( root ) {
+		global.addEventListener( 'error', ( error ) => wp.element.render( <ErrorScreen error={ error } /> ), root );
+
 		render( (
 			<Providers>
 				<Root appRoot={ root } />

```

I'll see if this actually works...

westonruter commented:
> I think we can accomplish this by adding an `error` event listener to the page that would then render the error screen.

Maybe this event listener should be removed once the app has successfully initialized. We wouldn't want errors caused by other random JS on the page to prevent the settings screen from rendering. 

Or rather, there could be a 5 second timeout after which if the app does not initialize, then allow whatever error(s) were caught by the `error` handler to be rendered. Upon initialization, the `error` event handler should then be removed, leaving only the error boundary error catcher in place.

delawski commented:
Those are some really good points. I'll dig into that 👍

delawski committed:
Add global error handler to settings page and onboarding wizard

delawski commented:
As suggested by @pierlon, in dea9fd4 we handle errors that are thrown outside of the error boundary by listening to the global `error` event.

The event listener is set up before the app is rendered. In the `render()` function's callback, the event listener is removed.

> We wouldn't want errors caused by other random JS on the page to prevent the settings screen from rendering.

@westonruter It seems that we can filter out "our own" errors by examining the `event.filename` string. It tells us where the error is coming from.

> Or rather, there could be a 5 second timeout after which if the app does not initialize, then allow whatever error(s) were caught by the error handler to be rendered. Upon initialization, the error event handler should then be removed, leaving only the error boundary error catcher in place.

I think that checking the `event.filename` combined with the error handler removal in the `render` function callback should do the trick.

---

In order to build and test out a global error handler as suggested in the previous comments, I first had to simulate an error similar the one in #6230 (a white screen of death).

<details>
<summary>
Check out my test steps
</summary>

First, I've added a dummy function to the compiled package `node_modules/@wordpress/i18n/build-types/index.d.ts`:

```diff
export { sprintf } from "./sprintf";
export * from "./create-i18n";
+ export function Foo() {
+ 	return null;
+ };
export { default as defaultI18n, setLocaleData, resetLocaleData, getLocaleData, subscribe, __, _x, _n, _nx, isRTL, hasTranslation } from "./default-i18n";
//# sourceMappingURL=index.d.ts.map
```

Thanks to that, webpack had no issues with importing that dummy function to our entry points:

```diff
- import { __ } from '@wordpress/i18n';
+ import { __, Foo } from '@wordpress/i18n';
```

Then, I've added a call to that function before the `render` function (so before we have the error boundary set up):

```diff
domReady( () => {
	const root = document.getElementById( 'amp-settings-root' );

+	Foo();

	if ( root ) {
		render( (
			<Providers>
				<Root appRoot={ root } />
			</Providers>
		), root );
	}
} );
```

Since in the WP admin the Core version of the `i18n` package is used, `Foo` becomes a missing dependency which leads to a white screen of death.

![Screenshot 2021-06-15 at 11 11 21](https://user-images.githubusercontent.com/478735/122026385-7120e080-cdca-11eb-9cea-6769400cfd83.png)

</details>

pierlon commented on assets/src/onboarding-wizard/index.js:
```diff
@@ -90,13 +95,24 @@ Providers.propTypes = {
+
+	const errorHandler = ( error ) => {
+		// Handle only own errors.
+		if ( error?.filename?.match( /amp-onboarding-wizard(\.min)?\.js/ ) ) {
```
This would be more appropriate since we're not making use of the matches:

```suggestion
		if ( /amp-onboarding-wizard(\.min)?\.js/.test( error?.filename ) ) {
```

pierlon commented on assets/src/onboarding-wizard/index.js:
```diff
@@ -90,13 +95,24 @@ Providers.propTypes = {
+		return;
 	}
+
+	const errorHandler = ( error ) => {
```
This is actually the [`ErrorEvent`](https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent), not the `error` object, so we could use a more appropriate name:

```suggestion
	const errorHandler = ( event ) => {
```

pierlon commented on assets/src/onboarding-wizard/index.js:
```diff
@@ -90,13 +95,24 @@ Providers.propTypes = {
+	const errorHandler = ( error ) => {
+		// Handle only own errors.
+		if ( error?.filename?.match( /amp-onboarding-wizard(\.min)?\.js/ ) ) {
+			render( <ErrorScreen error={ error } />, root );
```
The error event is passed here, instead of the error object, so this should actually be:

```suggestion
			render( <ErrorScreen error={ error.error } />, root );
```

pierlon commented on assets/src/onboarding-wizard/index.js:
```diff
@@ -90,13 +95,24 @@ Providers.propTypes = {
+			<SetupWizard closeLink={ CLOSE_LINK } finishLink={ FINISH_LINK } appRoot={ root } />
+		</Providers>,
+		root,
+		() => global.removeEventListener( 'error', errorHandler ),
```
According to the [React docs](https://reactjs.org/docs/react-dom.html#render) the callback is executed after the component is rendered or updated, which means the error boundary will not be handling errors until all renders are complete. Maybe we could remove the listener in the `Providers` component?

pierlon commented on assets/src/settings-page/index.js:
```diff
@@ -211,13 +217,24 @@ Root.propTypes = {
 	}
+
+	const errorHandler = ( error ) => {
+		// Handle only own errors.
```
Same comments above apply for this file as well.

pierlon committed:
Remove event listener as soon as possible

pierlon reviewed (approved):
I've made my suggestions in https://github.com/ampproject/amp-wp/pull/6320/commits/fa13a7f5ef7d2fe6dc93c6ca113357f4c1007868, feel free to modify if needed. That said, I think this is ready to ship :shipit:.

pierlon commented:
Oops one more thing:

The error stacktrace is not being displayed correctly in Firefox. Given this change:

```diff
diff --git a/assets/src/onboarding-wizard/index.js b/assets/src/onboarding-wizard/index.js
--- a/assets/src/onboarding-wizard/index.js	(revision fa13a7f5ef7d2fe6dc93c6ca113357f4c1007868)
+++ b/assets/src/onboarding-wizard/index.js	(date 1623874797950)
@@ -112,6 +112,8 @@
 
 	global.addEventListener( 'error', errorHandler );
 
+	foo();
+
 	render(
 		<Providers>
 			<SetupWizard closeLink={ CLOSE_LINK } finishLink={ FINISH_LINK } appRoot={ root } />
```

Here's what the error screen looks like in Firefox vs Chrome:

<table>
<tr>
	<td align="center">Firefox</td>
	<td align="center">Chrome</td>
<tr>
	<td><img width="1111" alt="image" src="https://user-images.githubusercontent.com/16200219/122290653-f935e180-cec1-11eb-9942-12a2dadf567f.png"></td>
	<td><img width="1199" alt="image" src="https://user-images.githubusercontent.com/16200219/122290580-e6bba800-cec1-11eb-9d79-15eb7023cfc2.png"></td>
</table>

Both browsers outputs `error.stack` differently, and in Firefox you can see that part of the stacktrace is incorrectly being rendered as an HTML comment.

pierlon committed:
Fix display of error stacktrace

pierlon commented:
Result after https://github.com/ampproject/amp-wp/pull/6320/commits/2a4b083eb11bdd802ede455a0fcda1a160ad3a5f:

<img width="1044" alt="image" src="https://user-images.githubusercontent.com/16200219/122291651-0d2e1300-cec3-11eb-86c4-3ad6c714f362.png">

delawski committed:
Update snapshot

delawski commented:
@pierlon Thank you for the feedback and for addressing it right away. I've updated the `ErrorScreen` snapshot and pushed it so that the unit tests should now pass.

westonruter commented:
I just tried this without the fix from https://github.com/ampproject/amp-wp/pull/6396:

Chrome | Firefox
---------|---------
![image](https://user-images.githubusercontent.com/134745/122477411-8473ad80-cf7c-11eb-8c4f-e5acee4d878a.png) | ![image](https://user-images.githubusercontent.com/134745/122477602-dae0ec00-cf7c-11eb-91c6-59a64433b139.png)

It seems like in Firefox the error message is not included in the details.

pierlon commented:
Maybe the JS sourcemap isn't being detected properly in Firefox?

westonruter commented:
I'm talking about the contents of the `details`. In Chrome the "TypeError: Cannot read property 'edit_link' of undefined" message is part of the details. In Firefox it is not. This means the error message is not copied to the clipboard in Firefox.

pierlon commented:
What about something like:

```diff
diff --git a/assets/src/components/error-screen/index.js b/assets/src/components/error-screen/index.js
--- a/assets/src/components/error-screen/index.js	(revision 2a4b083eb11bdd802ede455a0fcda1a160ad3a5f)
+++ b/assets/src/components/error-screen/index.js	(date 1623968168232)
@@ -27,6 +27,7 @@
  */
 export function ErrorScreen( { error, finishLinkLabel, finishLinkUrl, title } ) {
 	const [ hasCopied, setHasCopied ] = useState( false );
+	const { message, stack } = error;
 
 	return (
 		<div className="error-screen-container">
@@ -37,21 +38,21 @@
 
 				{ /* dangerouslySetInnerHTML reason: WordPress sometimes sends back HTML in error messages. */ }
 				<p dangerouslySetInnerHTML={ {
-					__html: error.message || __( 'There was an error loading the page.', 'amp' ),
+					__html: message || __( 'There was an error loading the page.', 'amp' ),
 				} } />
 
-				{ error?.stack && (
+				{ stack && (
 					<details>
 						<summary>
 							{ __( 'Details', 'amp' ) }
 						</summary>
 						<pre>
-							{ error.stack }
+							{ stack }
 						</pre>
 						<ClipboardButton
 							isSmall={ true }
 							isSecondary={ true }
-							text={ error.stack }
+							text={ JSON.stringify( { message, stack }, null, 2 ) }
 							onCopy={ () => setHasCopied( true ) }
 							onFinishCopy={ () => setHasCopied( false ) }
 						>
```

Which would copy the error message and stack.

westonruter commented:
Yeah, that looks good.

pierlon committed:
Copy error stacktrace along with the message

pierlon commented:
Done in https://github.com/ampproject/amp-wp/pull/6320/commits/94d183e6b3c7dbe69065ff2e5840925fb7479b6f.

westonruter commented:
Text on clipboard for Chrome:

```json
{
  "message": "Cannot read property 'edit_link' of undefined",
  "stack": "TypeError: Cannot read property 'edit_link' of undefined\n    at l (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:27:100)\n    at Ui (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:72225)\n    at fc (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:118455)\n    at ss (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:111004)\n    at os (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:110929)\n    at Qc (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:107959)\n    at https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59614\n    at t.unstable_runWithPriority (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:134838)\n    at Pr (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59323)\n    at Ir (https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59559)"
}
```

Text on clipboard for Firefox:

```json
{
  "message": "t.user is undefined",
  "stack": "l@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:27:59\nUi@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:72225\nfc@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:118455\nss@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:111006\nos@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:110929\nQc@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:107959\nIr/<@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59614\nt.unstable_runWithPriority@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:134838\nPr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59323\nIr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59561\nFr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59494\nW@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:128628\nVt@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:37019\nEventListener.handleEvent*Ut@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:36943\nIt@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:36745\nMt@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:32339\non@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:40036\nJo@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:90431\nls@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:111187\nss@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:111076\nos@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:110929\nQc@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:107959\nIr/<@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59614\nt.unstable_runWithPriority@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:134838\nPr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59323\nIr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59561\nFr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59494\nVc@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:104748\npo@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:76571\nl/</<@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:66159\nasync*l/<@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:66165\nac@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:97499\nfs@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:116574\nt.unstable_runWithPriority@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:134838\nPr@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:59323\nps@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:116333\nms/<@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:116244\nB@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:133778\nO.port1.onmessage@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:132534\nEventHandlerNonNull*@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:132458\nn@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:110\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:130961\nn@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:110\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:14371\nn@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:110\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:123654\nn@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:110\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:3671\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:59:6889\nn@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:110\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:902\n@https://wordpressdev.lndo.site/content/plugins/amp/assets/js/amp-settings.js?ver=868573821750c3d8c887152600dbc6fe:1:913\n"
}
```

westonruter commented on assets/src/settings-page/index.js:
```diff
@@ -211,13 +221,23 @@ Root.propTypes = {
+
+	errorHandler = ( event ) => {
+		// Handle only own errors.
+		if ( /amp-settings(\.min)?\.js/.test( event?.filename ) ) {
```
```suggestion
		if ( event.filename && /amp-settings(\.min)?\.js/.test( event.filename ) ) {
```

westonruter commented on assets/src/onboarding-wizard/index.js:
```diff
@@ -90,13 +99,23 @@ Providers.propTypes = {
+
+	errorHandler = ( event ) => {
+		// Handle only own errors.
+		if ( /amp-onboarding-wizard(\.min)?\.js/.test( event?.filename ) ) {
```
I think this conditional makes more sense like so:

```suggestion
		if ( event.filename && /amp-onboarding-wizard(\.min)?\.js/.test( event.filename ) ) {
```

In other words, it doesn't make sense to possibly pass `null` to the `test()` method.

westonruter commented on assets/src/components/clipboard-button/index.js:
```diff
@@ -0,0 +1,66 @@
+
+const TIMEOUT = 4000;
+
+// Migrated from @wordpress/components.
```
```suggestion
// Adapted from @wordpress/components: <https://github.com/WordPress/gutenberg/blob/3c00d85/packages/components/src/clipboard-button/index.js#L18-L69>.
```

westonruter commented on assets/src/components/clipboard-button/index.js:
```diff
@@ -0,0 +1,66 @@
+const TIMEOUT = 4000;
+
+// Migrated from @wordpress/components.
+export default function ClipboardButton( {
```
I suppose `wp.compose.useCopyToClipboard` not being used to have more compatibility with older versions of Gutenberg?

delawski replied:
Our `ClipboardButton` component is actually [utlizing](https://github.com/ampproject/amp-wp/pull/6320/files#diff-a47b24c70cb0eda5614a90acdb3a84ce27f313507d2a03d76ade03e5a6f2e626R24) the `wp.compose.useCopyToClipboard` hook.

I'm not entirely sure why Gutenberg is phasing out the `ClipboardButton` component. I suppose they want to have more freedom in how the button looks and behaves depending on the location in the UI.

westonruter commented on assets/src/components/error-screen/index.js:
```diff
@@ -12,34 +13,61 @@ import PropTypes from 'prop-types';
+
+				{ /* dangerouslySetInnerHTML reason: WordPress sometimes sends back HTML in error messages. */ }
+				<p dangerouslySetInnerHTML={ {
+					__html: message || __( 'There was an error loading the page.', 'amp' ),
```
This would probably be good to have a link to the support forum: https://wordpress.org/support/plugin/amp/

Something like this:

```suggestion
					__html: ( message || __( 'There was an error loading the page.', 'amp' ) ) + ' ' + __( 'Please submit details our <a href="https://wordpress.org/support/plugin/amp/" target="_blank" rel="noreferrer noopener">support forum</a>.', 'amp' ),
```

Naturally perhaps `sprintf()` should be used instead.

delawski replied:
That's a good idea. In 965e9ec a link to the support forum has been added. It's a separate P tag since I'm not sure if every JS error message ends with a full stop or not. With two paragraph tags, it doesn't really matter that much:

![Screenshot 2021-06-21 at 12 51 54](https://user-images.githubusercontent.com/478735/122750976-77fa9800-d28f-11eb-9242-68c7ed0a3737.png)

westonruter reviewed (changes requested):
I tried editing `api-fetch.js` to throw an error on its first line, and the result I got on the settings screen was:

<img width="873" alt="Screen Shot 2021-06-17 at 16 47 07" src="https://user-images.githubusercontent.com/134745/122485447-c8ba7a00-cf8b-11eb-9392-1d593e47bc99.png">

Can this case of an error in a dependency be caught as well?

delawski committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Add support forum link to error screen message

delawski commented:
> I tried editing `api-fetch.js` to throw an error on its first line, and the result I got on the settings screen was: [...]
Can this case of an error in a dependency be caught as well?

@westonruter I think that handling this kind of error would be overkill.

We normally use the Core `wp.apiFetch` package and polyfill it only on WordPress 4.9 and older versions. The polyfill itself is a very thin wrapper around the `@wordpress/api-fetch` package. It's enqueued before our JS entry point so we don't have our regular error event handler in place at this point. We would need a separate event handler, ideally outside the polyfill itself, so that it would catch potential errors coming from the `@wordpress/api-fetch` package.

Also, the `polyfill/api-fetch.js` entry point does not depend on React right now while the error screen is React-based. So we would either have to recreate the error screen in vanilla JS or pull in the `@wordpress/element` package into the polyfill. It changes the bundle size drastically:

| Current polyfill filesize | Polyfill with `@wordpress/element` |
| --- | --- |
| <img width="163" alt="Screenshot 2021-06-21 at 14 42 48" src="https://user-images.githubusercontent.com/478735/122764836-2ad2f200-d2a0-11eb-9a5e-b7bb47dea07c.png"> | <img width="155" alt="Screenshot 2021-06-21 at 14 42 35" src="https://user-images.githubusercontent.com/478735/122764855-2eff0f80-d2a0-11eb-8fe8-f8c771726933.png"> |

I'm wondering about another approach. Maybe we could handle the pre-main-entry point errors, store them, and then read on the initial render. If there would be any errors stored, we would show the error screen (within our main entry point). I'm just not sure, though, if we want to go that far.

I'm happy to discuss different approaches, too.

delawski committed:
Update snapshot

westonruter commented:
> We normally use the Core `wp.apiFetch` package and polyfill it only on WordPress 4.9 and older versions. The polyfill itself is a very thin wrapper around the `@wordpress/api-fetch` package.

Actually, I was just using api-fetch as an example. What an error happens in any of the dependencies? Maybe it's still overkill. Or what if one of the dependencies fails to load because of a network problem?

> I'm wondering about another approach. Maybe we could handle the pre-main-entry point errors, store them, and then read on the initial render. If there would be any errors stored, we would show the error screen (within our main entry point). I'm just not sure, though, if we want to go that far.

Perhaps a middle-ground, instead of trying to capture the errors to then to just show a generic error message with some instructions for how to 

One idea would be to have a CSS animation that shows an error message after 5 seconds?

For example, inside of `#amp-settings-root` there could be something like:

```html
<div id="amp-settings-root">
    <div id="amp-loading-failure">Oops! Something went wrong. Please open your browser console to see the error messages and share them on the <a href="https://wordpress.org/support/plugin/amp/">support forum</a>.
    <noscript>
        <p>You must have JavaScript enabled to use the AMP Settings screen.</p>
    </noscript>
</div>
```

(Note also the `noscript` message.)

This could be accompanied by CSS like the following:

```css
#amp-settings-root > #amp-loading-failure {
    visibility: hidden;
    animation: amp-wp-show-error 5s steps(1, end) 0s 1 normal both
}
@keyframes amp-wp-show-error {
	from {
		visibility: hidden;
	}
	to {
		visibility: visible;
	}
}
```

If the JS fails to load, then the `#amp-settings-root` won't get emptied out and so the error message will be shown after 5 seconds.

delawski committed:
Handle async errors within error boundary

delawski committed:
Ensure error message does not overflow the container

delawski committed:
Handle errors thrown when JavaScript is loaded

delawski commented:
> If the JS fails to load, then the #amp-settings-root won't get emptied out and so the error message will be shown after 5 seconds.

That's actually a very good idea! I've added a loading error handling in d49f001.

I've also noticed that async errors thrown in the context providers are not properly handled by the error boundary. This should be now fixed with f22c055.

delawski committed:
Use `render` instead of reserved `print` keyword

westonruter committed:
Fix OnboardingWizardSubmenuPageTest::test_render

westonruter commented:
(Rebased against `develop`.)

westonruter committed:
Show pre-loading spinner and show error immediately when JS disabled

westonruter committed:
Hide validation counts in admin menu when JS is disabled

westonruter committed:
Scope box-sizing style to .error-screen-container

westonruter committed:
Increase timeout to show error screen for sake of slow connections

westonruter commented:
@delawski @pierlon I made a few more additional improvements:

1. I increased the timeout to show the uncaught error message from 5 to 30 seconds, as this accounts for a slow connections.
1. When JS is disabled the no-JS message is now shown immediately without having to wait for the animation timeout.
1. When JS is disabled, I also hid the spinners for the validation counts since they will never be populated.
1. I added a `spinner.gif` to show in the HTML before the `Loading` JS component is rendered. This ensures that there is a spinner showing on the screen to let them know to wait (for up to 30 seconds).
1. I scoped the `box-sizing` property in 3d636b3 because it was causing the validation error count loading animation in the admin menu to become mistyled.

## Successful Load

Notice how there are two spinners: a pre-loading spinner and a loading spinner.

https://user-images.githubusercontent.com/134745/123009314-5245bf80-d371-11eb-9b0d-2bcb038334af.mov

## Failure due to JS disabled

<img width="1003" alt="amp-settings-load-failure-no-js" src="https://user-images.githubusercontent.com/134745/123009353-5ffb4500-d371-11eb-84a1-06c00691d46e.png">

## Failure with caught error

Here I tested by throwing an error at the top of `api-fetch.js`:

https://user-images.githubusercontent.com/134745/123009382-6ee1f780-d371-11eb-9a6b-936335cc5b24.mov

## Failure with uncaught error

Here I tested by throwing an error at the top of `lodash.js` (note this takes 30 seconds):

https://user-images.githubusercontent.com/134745/123009612-de57e700-d371-11eb-82c8-98684302fc84.mov

westonruter committed:
Use service for dependency injection and add test

westonruter committed:
Fix translation function

westonruter committed:
Add link to clipboard-button component with the specific commit

westonruter reviewed (approved):
Great work! 

@delawski @pierlon please review the additional changes and I'll merge tomorrow once all verified.

delawski committed:
Use the same spinner animation while pre-loading

delawski commented:
@westonruter Thank you for the improvements. They all look good to me.

The only thing that I changed in 26c1206 is the loading spinner. Right now it looks exactly the same as in the React app so that it's hard to notice when they get swapped:

https://user-images.githubusercontent.com/478735/123121582-90141800-d445-11eb-8ada-24351e04a682.mp4

delawski committed:
Update unit test

pierlon commented on src/Admin/OnboardingWizardSubmenuPage.php:
```diff
@@ -60,17 +61,26 @@ final class OnboardingWizardSubmenuPage implements Delayed, Registerable, Servic
 	private $rest_preloader;
 
+	/**
+	 * RESTPreloader instance.
```
```suggestion
	 * LoadingError instance.
```

pierlon commented on src/LoadingError.php:
```diff
@@ -0,0 +1,147 @@
+
+			#amp-pre-loading-spinner {
+				visibility: visible;
+				animation: amp-wp-hide-element 30s steps(1, end) 0s 1 normal both; /* This could probably reuse amp-wp-show-element if reversed. */
```
I'm not sure how a slow connection would come into play here, as the page would have already loaded and the necessary JS would have either executed or not by then. A 30s wait time seems to be overkill here.

westonruter replied:
I was testing with Fast 3G and the JS files were taking over 30 seconds to load. Granted, I was testing with the development versions, so it was 8MB (😱). I'll test again with the production version instead and adjust the timeout down.

westonruter replied:
With Fast 3G, the settings page takes 10 seconds to load. With Slow 3G it takes 36 seconds to load. This is with the production build, where `amp-settings.js` is 875KB. So I think 30 seconds is not actually far off.

pierlon replied:
Ah I see. That makes sense.

pierlon replied:
What an ingenious solution!

westonruter committed:
Fix phpdoc comment

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

westonruter commented:
> The only thing that I changed in [26c1206](https://github.com/ampproject/amp-wp/commit/26c12061490737696199e4f332004f0520e9ed99) is the loading spinner. Right now it looks exactly the same as in the React app so that it's hard to notice when they get swapped:

Nice improvement. I do notice the second spinner shifts a tiny bit when it loads, but otherwise it's better as you have it now since the `loading.gif` doesn't load unnecessarily.

pierlon reviewed (approved):
This looks ready to ship. Great work @delawski and @westonruter!

westonruter commented on src/LoadingError.php:
```diff
@@ -0,0 +1,147 @@
+				width: 18px;
+				height: 18px;
+				opacity: 0.7;
+				margin: 5px 11px 0;
```
@delawski why the `margin`? This is not present in the `Loading` component and that's causing the slight layout shift when the static spinner is replaced with the React one. I think this line simply should be removed:

```suggestion
```

westonruter replied:
Seems this was copied from `spinner-stylers.js`:

![image](https://user-images.githubusercontent.com/134745/123167953-29cecb80-d42c-11eb-905c-790b643dadbc.png)

Buy as you can see the `margin` is overridden by `assets/src/components/loading/style.css`:

https://github.com/ampproject/amp-wp/blob/a958b4356074983d589e997bb518922aea8dd76f/assets/src/components/loading/style.css#L7-L9

So yeah, I'm just going to remove the `margin` to make it consistent.

westonruter committed:
Remove extraneous margin from spinner
